nullaway/src/main/java/com/uber/nullaway/handlers/EnsuresNonnullHandler.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.handlers;
+
+import static com.google.errorprone.BugCheckerInfo.buildDescriptionFromChecker;
+
+import com.google.common.base.Preconditions;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.api.JavacTrees;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.util.Context;
+import com.uber.nullaway.ErrorMessage;
+import com.uber.nullaway.NullAway;
+import com.uber.nullaway.Nullness;
+import com.uber.nullaway.dataflow.AccessPath;
+import com.uber.nullaway.dataflow.AccessPathNullnessPropagation;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+
+public class EnsuresNonnullHandler extends BaseNoOpHandler {
+
+  private static final String annotName = "com.uber.nullaway.qual.EnsuresNonnull";
+  private static final String thisNotation = "this.";
+
+  private @Nullable NullAway analysis;
+  private @Nullable VisitorState state;
+
+  @Override
+  public void onMatchTopLevelClass(
+      NullAway analysis, ClassTree tree, VisitorState state, Symbol.ClassSymbol classSymbol) {
+    this.analysis = analysis;
+    this.state = state;
+  }
+
+  @Override
+  public void onMatchMethod(
```

Comment:
I think we need to check here for cases where this method overrides a superclass method, and the superclass method has `@EnsuresNonNull`?  In such cases we need the overriding method to adhere to the `@EnsuresNonNull` spec from the superclass method, right?

Reply:
Hi Manu, thank you very much for your comments, I addressed them all. I believe this is ready for another review.

Please find the latest updates below from your last review:

1. Addressed all comments.
2. Added the support for selecting multiple class fields in both annotations.
3. Fixed the bug where the user might choose a field from super classes fields.
4. Fixed the bug where a method might not be annotated as ```@EnsuresNonNull``` but its superclass does, and it was not updating the ```NullnessStore``` when the method has been invoked.
5. Added the inheritance support for annotations:
   * ```@EnsuresNonNull```: All sub methods, can only add new fields to the set of all fields mentioned as parameter in super class ```@EnsuresNonNull``` annotation. All selected class fields including super methods selected fields must be written explicitly as parameter in the sub methods ```@EnsuresNonNull``` annotation. If a method is not annotated as ```@EnsuresNonNull```, still needs to guarantee all selected fields from super method are ```@NonNull``` at exit point if exists.
   * ```@RequiresNonNull```: All sub methods, can only remove fields from the set of all fields mentioned as parameter in super class ```@RequiresNonNull``` annotation. If a method is not annotated as ```@RequiresNonNull```, it will not be processed and no check will take place when it has been invoked, even if the super method is annotated with ```@RequiresNonNull``` annotation.
6. Refactored the code since ```RequiresNonNullHandler``` and ```EnsuresNonNullHandler``` had a lot in common.